tough/src/http.rs

```diff
@@ -105,19 +104,12 @@ impl Read for RetryRead {
                 return Err(retry_err);
             }
             self.retry_state.increment(&self.settings);
+            self.err_if_no_range_support(retry_err)?;
             // wait, then retry the request (with a range header).
             std::thread::sleep(self.retry_state.wait);
-            if !self.supports_range() {
-                // we cannot send a byte range request to this server, so return the error
-                error!(
-                    "an error occurred and we cannot retry because the server \
-                    does not support range requests '{}': {:?}",
-                    self.url, retry_err
-                );
-                return Err(retry_err);
-            }
             let new_retry_read =
-                fetch_with_retries(&mut self.retry_state, &self.settings, &self.url)?;
+                fetch_with_retries(&mut self.retry_state, &self.settings, &self.url)
+                    .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e))?;
```

Comment:
The implementation is confusing by nature, it might actually require a meeting to try and go through it. Basically what happens is:
- `Read` is a trait, it returns a `std::io::Error/Result`
- `fetch` returns a transport error.

Let's walk through a case where a broken-pipe occurs while reading the request body. It goes like this:

- `fetch` returns Ok(Read)
- User calls Read
- Error occurs in the inner Read object, this will be an io::Error
- We do retries, which means we are using fetch again, this returns a TransportError/Result.
- If we try fetch and fail, run out of retries etc, now we are holding a TransportError... but:
- We are inside of a Read object's `read` function so we need to wrap it up in std::io::Error/Result

Reply:
P.S. The implementation of `From<HttpError> for std::io::Error` was completely unused and probably added to the confusion, I have removed it.